### ISSUE TITLE
feat: tasks created by extensions should be cleared from TaskManager when extension are disabled/removed

### ISSUE THREAD

**Issue by axel7083** [containers/podman-desktop#5804]
### Is your enhancement related to a problem? Please describe

When an extension is created a lot of tasks, some stuff might go wrong, and disabling/enabling the extension does not fix the issue as some ghost never ending tasks are displayed displayed

### Describe the solution you'd like

We can add to any tasks created an optional `extensionId` to track the creator of the Tasks, so if an extension is disabled we clear its pending tasks.

**Comment by deboer-tim** [comment]
Personally I don't like this approach - e.g. if I did a podman build and then removed Podman extension I still did the podman build. If I'm just doing an update (which involves removing the old extension and installing the new one) I'm assuming we wouldn't want to remove them either. Plus, sometimes in dev there's a forever-task and I'd like to just remove it without disabling/enabling.

I'd rather there is just a way to correctly kill off bad/never-ending tasks.

**Comment by axel7083** [comment]
@deboer-tim I think that the tasks should be cancelled/stopped if the extension is disabled. if I am deleting a program in my computer, I am not expecting tasks/actions they start to still be running, this action is like a stop everything.

If I am making an update while an extension is building or doing some background tasks, I think it makes sense for those tasks to be stopped, as I never want some old extension code run in parallel of new extension.

**Comment by deboer-tim** [comment]
That doesn't sound like the original description to me :), but if the intent is to just make sure any running or ghost task is just stopped/cancelled when the extension are stopped (and you can clean them up using the regular clear button), 👍🏼 from me.

**Comment by axel7083** [comment]
yeah I mean, I used the technical terms in the title, and I probably shouldn't as it is missleading probably on the real purpose

**Comment by benoitf** [comment]
arriving late in the party but we don't need extensionId to cancel tasks. 

I've, we won't do something like task manager.cleanupTasksForExtension(extensionId)


Usually we are/should use the disposable pattern
So you call dispose() to remove objects and you push disposables to the extension context disposables array

When stopping extensions we dispose objects if not already disposed

**Comment by axel7083** [comment]
> arriving late in the party but we don't need extensionId to cancel tasks.
> 
> I've, we won't do something like task manager.cleanupTasksForExtension(extensionId)
> 
> Usually we are/should use the disposable pattern So you call dispose() to remove objects and you push disposables to the extension context disposables array
> 
> When stopping extensions we dispose objects if not already disposed

Yeah that's the theory if everything is fine and working, I really like the Disposable pattern but that requires the extension to properly manage it. And about the tasks system, when I am using a `windows.withProgress` I am not getting a Disposable, I am getting a promise. 

But sadly what do we do if the extension never dispose them, because they crash or something, what's  the user's option to clear ? Reboot the full Podman-Desktop application.

If rebooting the extension does not fix it, user might be frustated to have a lot of tasks (probably still running the indeterminate animation) accumulate, we might want give users some abilities to troubleshoots and fix things themselves. 

Let's take some inspiration in Android, applications can create Notifications, notification lifecycle are managed by the application, they can update, clear or change the state of the notification, but Android keep a control over them, if the application is being uninstalled, it is removing all the notifications and processes remaining. 

I have a feeling we should have some cleanup mechanism on top of the disposable pattern to keep things nice and clean.

**Comment by deboer-tim** [comment]
If we were going the way of Android we'd have to redo a _lot_ of API like `registerCommand` to make sure every path leads to a disposable getting pushed - but I think VS Code shows this path can work well too and it has been doing fine for us.

IMHO the trick is making it easy and natural. Most new extensions are going to register commands with some form of:
```
extensionContext.subscriptions.push(
    extensionApi.commands.registerCommand( ...
```
because it's simple and what the examples show, but none of them are going to the same level of defensive/cleanup code for `windows.withProgress` because, well, we don't know ourselves. It might need new API if the extension can't reasonably handle it itself, but if so Florent is right that disposable pattern is the best fit/most consistent here.

**Comment by benoitf** [comment]
about disposables we should automatically register them in addition of having the extensions pushing it as well.

All registration of commands could be cleaned automatically ( when we create command we push the disposable to the list of disposables to clean when unregistering the extension)

About the crash of an extension it could handle it as well ( that said I'm not sure what you call a crash)

**Comment by axel7083** [comment]
> about disposables we should automatically register them in addition of having the extensions pushing it as well.
> 
> All registration of commands could be cleaned automatically ( when we create command we push the disposable to the list of disposables to clean when unregistering the extension)
> 

I think that we could we _could_ register automatically the disposable as well, but then would we recommend the extension to do it too?

For the `windows.withProgress` we could create our own disposable internally, not exposing it to the extension as it is not suppose to care about it being disposed.

> About the crash of an extension it could handle it as well ( that said I'm not sure what you call a crash)

Simply a case where an error is not properly handled, and an exception is raised in a specific manner than the task/function given to the withProgress is never resolved/rejected

**Comment by axel7083** [comment]
This is probably covered by https://github.com/containers/podman-desktop/pull/7612
